src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -173,14 +173,14 @@ internal JsonConverter<TTarget> CreateCastingConverter<TTarget>()
         /// </summary>
         internal bool IsInternalConverterForNumberType { get; init; }
 
-        internal static bool ShouldFlush(ref WriteStack state)
+        internal static bool ShouldFlush(ref WriteStack state, Utf8JsonWriter writer)
         {
             Debug.Assert(state.FlushThreshold == 0 || (state.PipeWriter is { CanGetUnflushedBytes: true }),
                 "ShouldFlush should only be called by resumable serializers, all of which use the PipeWriter abstraction with CanGetUnflushedBytes == true.");
             // If surpassed flush threshold then return true which will flush stream.
             if (state.PipeWriter is { } pipeWriter)
             {
-                return state.FlushThreshold > 0 && pipeWriter.UnflushedBytes > state.FlushThreshold;
+                return state.FlushThreshold > 0 && pipeWriter.UnflushedBytes > state.FlushThreshold - writer.BytesPending;
```

Comment:
Do we need to update the doc comment for `Utf8JsonWriter.BytesPending`? Or should we change the behavior to match it?

```csharp
        /// <summary>
        /// Returns the amount of bytes written by the <see cref="[Utf8JsonWriter](https://source.dot.net/System.Text.Json/P/3245a5b5ce1d7e64.html#3245a5b5ce1d7e64)"/> so far
        /// that have not yet been flushed to the output and committed.
        /// </summary>
```

This doc comment is correct for `Stream` but not for `PipeWriter`. In the case of a `PipeWriter`, the number of unflushed bytes is the more interesting than the number of uncommitted bytes which is pretty hard to explain without going into the intricacies of `IBufferWriter`, `PipeWriter` and `Stream` APIs and how they compare and contrast.

Especially as public API, `BytesCommited` and `BytesPending` meaning unadvanced instead of unflushed provides very little value. Even if bytes are "pending" and not "committed" because `Advance` hasn't been called yet, I cannot think of a single use case where that distinction matters. It's not like developers can do anything with the uncommitted bytes referenced `Utf8JsonWriter._memory`.

What consumers of the public `Utf8JsonWriter` type might care about is whether they should flush their `PipeWriter` or `Stream` just like our `Serialize`/`SerializeAsync` methods do. And in that case, it would be better if `BytesPending` represented unflushed bytes.

As it stands, it's odd that given a `Stream`, `Utf8JsonWriter.FlushAsync` actually flushes the Stream but merely advances the `PipeWriter`. I know that's because `Utf8JsonWriter` only sees the `PipeWriter` as an `IBufferWriter` which doesn't have the ability to flush, but this feels wrong given that we're trying to add proper `PipeWriter` support to S.T.J in .NET 9. Now, at least in the case of the `Utf8JsonWriter` API's, `PipeWriter` support feels very second class.

Reply:
That sounds like we should add PipeWriter support to Utf8JsonWriter as well right?

Reply:
What do you suggest that wouldn't be a breaking change?